#  Pandas project  - Jacobo Pintos

En este documento buscaremos limpiar el dataset descargado para poder en el siguiente documento realizar un ánalisis sobre los datos extraídos. Para la realización deberemos importar las siguientes librerías:
    - Pandas
    - Numpy
    - Re

In [1]:
import pandas as pd
import numpy as np
import re

## 1. Dataset extraction and initial filtering

Immportamos el documento y le cambiamos la codificación para que se pueda leer en pandas por medio de la siguiente función. Nombraremos el dataset inicialmente como "data".

In [2]:
data = pd.read_csv('GSAF5.csv', encoding='ISO/8859/1')

Vemos si la importación ha sido correcta y visualizamos una muestra de 20 valores de los datos extraídos.

In [3]:
data.head(10)

Case Number       Date  Year        Type        Country  \
0  2016.09.18.c  18-Sep-16  2016  Unprovoked            USA   
1  2016.09.18.b  18-Sep-16  2016  Unprovoked            USA   
2  2016.09.18.a  18-Sep-16  2016  Unprovoked            USA   
3    2016.09.17  17-Sep-16  2016  Unprovoked      AUSTRALIA   
4    2016.09.15  16-Sep-16  2016  Unprovoked      AUSTRALIA   
5  2016.09.15.R  15-Sep-16  2016        Boat      AUSTRALIA   
6    2016.09.11  11-Sep-16  2016  Unprovoked            USA   
7    2016.09.07  07-Sep-16  2016  Unprovoked            USA   
8    2016.09.06  06-Sep-16  2016  Unprovoked  NEW CALEDONIA   
9  2016.09.05.b  05-Sep-16  2016  Unprovoked            USA   

                Area                                         Location  \
0            Florida                 New Smyrna Beach, Volusia County   
1            Florida                 New Smyrna Beach, Volusia County   
2            Florida                 New Smyrna Beach, Volusia County   
3           Victoria                                 Thirteenth Beach   
4           Victoria                                      Bells Beach   
5  Western Australia                                          Bunbury   
6            Florida                    Ponte Vedra, St. Johns County   
7             Hawaii                                     Makaha, Oahu   
8     North Province                                           Koumac   
9     South Carolina  Kingston Plantation, Myrtle Beach, Horry County   

          Activity                    Name Sex   ...          Species   \
0          Surfing                    male    M  ...               NaN   
1          Surfing          Chucky Luciano    M  ...               NaN   
2          Surfing                    male    M  ...               NaN   
3          Surfing         Rory Angiolella    M  ...               NaN   
4          Surfing                    male    M  ...         2 m shark   
5          Fishing  Occupant: Ben Stratton  NaN  ...               NaN   
6           Wading                    male    M  ...    3' to 4' shark   
7         Swimming                  female    F  ...  Tiger shark, 10?   
8     Kite surfing            David Jewell    M  ...               NaN   
9  Boogie boarding          Rylie Williams    F  ...               NaN   

        Investigator or Source                        pdf  \
0  Orlando Sentinel, 9/19/2016       2016.09.18.c-NSB.pdf   
1  Orlando Sentinel, 9/19/2016   2016.09.18.b-Luciano.pdf   
2  Orlando Sentinel, 9/19/2016       2016.09.18.a-NSB.pdf   
3           The Age, 9/18/2016  2016.09.17-Angiolella.pdf   
4           The Age, 9/16/2016  2016.09.16-BellsBeach.pdf   
5   West Australian, 9/15/2016      2016.09.15.R-boat.pdf   
6          News4Jax, 9/11/2016  2016.09.11-PonteVedra.pdf   
7    Hawaii News Now, 9/7/2016        2016.09.07-Oahu.pdf   
8       TVANouvelles, 9/6/2016      2016.09.06-Jewell.pdf   
9            C. Creswell, GSAF  2016.09.05.b-Williams.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   
5  http://sharkattackfile.net/spreadsheets/pdf_di...   
6  http://sharkattackfile.net/spreadsheets/pdf_di...   
7  http://sharkattackfile.net/spreadsheets/pdf_di...   
8  http://sharkattackfile.net/spreadsheets/pdf_di...   
9  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.c   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.b   
2  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.a   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2016.09.17   
4  http://sharkattackfile.net/spreadsheets/pdf_di...

Debemos empezar a analizar las columnas que requeriremos en el analisis y para ello visualizaremos los nombres de todas las columnas del dataset. 

In [4]:
data.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

Renombramos las columnas que requeriremos más tarde y tienen los nombres mal puestos.

In [5]:
data.rename(columns={'Sex ':'Sex'}, inplace=True)

Una vez renombradas las columnas crearemos una nueva base de datos con las columnas que necesitaremos a posteriori para el ánalisis, despreciando aquellas que no nos interesan. Renombramos el dataset con el nombre "sharks" para diferenciarlo del anterior.

In [6]:
colnames=['Year', 'Type', 'Country', 'Area', 'Sex', 'Age', "Fatal (Y/N)"]
sharks = pd.DataFrame(data, columns=colnames)

Una vez creado el sub data set vamos a ver como son los datos que tenemos del mismo y cuantas entradas nulas existen.

In [7]:
sharks.info()

null_cols = sharks.isnull().sum()
null_cols[null_cols > 0]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5992 entries, 0 to 5991
Data columns (total 7 columns):
Year           5992 non-null int64
Type           5992 non-null object
Country        5949 non-null object
Area           5590 non-null object
Sex            5425 non-null object
Age            3311 non-null object
Fatal (Y/N)    5973 non-null object
dtypes: int64(1), object(6)
memory usage: 327.8+ KB


Country          43
Area            402
Sex             567
Age            2681
Fatal (Y/N)      19
dtype: int64

Limitamos el dataset ya que el análisis se basará en ataques desde 1950 y en USA por tanto filtramos el dataset quitando todos los datos que sean de fuera de USA y anteriores a 1950. Una vez realizado vemos cúantos datos nos quedan.

In [8]:
sharks = sharks.drop(sharks[sharks['Year']<1950].index)
sharks = sharks.drop(sharks[sharks['Country']!='USA'].index)

In [9]:
sharks.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1780 entries, 0 to 4264
Data columns (total 7 columns):
Year           1780 non-null int64
Type           1780 non-null object
Country        1780 non-null object
Area           1779 non-null object
Sex            1699 non-null object
Age            1370 non-null object
Fatal (Y/N)    1780 non-null object
dtypes: int64(1), object(6)
memory usage: 111.2+ KB


Tras realizar esta función nos encontramos con una mejor cantidad de datos nulos.Previamente la columna Age contenía un 44% de valores nulos vs el 23% actual. 

## 2. Column Type cleaning

Debido al filtrado de años conocemos que todos los valores de la columna "Year" son únicos y enteros por tanto procedemos a analizar la tipología de la columna "Type"

In [10]:
sharks['Type'].value_counts()

Unprovoked      1427
Provoked         150
Invalid          145
Boat              38
Sea Disaster      18
Boating            2
Name: Type, dtype: int64

Podemos ver un total de 145 registros "Invalid" y además de otrás categorías que quiero saber si serán interesantes para mi análisis: 

    - "Boat"
    - "Sea Disaster"
    - "Boating" 
    
Debo anaizarlas individualmente para poder decidir si debo eliminarlas o no.

In [11]:
test_typeinvalid= sharks.drop(sharks[sharks['Type']!='Invalid'].index)
test_typeinvalid.head()

Year     Type Country            Area  Sex  Age Fatal (Y/N)
50   2016  Invalid     USA  South Carolina    M   27           N
75   2016  Invalid     USA         Florida    M  NaN           N
81   2016  Invalid     USA      California  NaN  NaN           N
116  2015  Invalid     USA         Florida    F    9           N
126  2015  Invalid     USA          Hawaii    M   32           N

Podemos observar como la mayoría de los "Invalid" no representan ataques de tiburón. Estos se dividen en datos no confirmados o cuestionable presencia de tiburón en el suceso. Deben eliminarse.

In [12]:
test_seadisaster= sharks.drop(sharks[sharks['Type']!='Sea Disaster'].index)
test_seadisaster.head()

Year          Type Country            Area Sex             Age  \
935   2009  Sea Disaster     USA     Puerto Rico   M             NaN   
1366  2005  Sea Disaster     USA  South Carolina   M           Teens   
2197  1994  Sea Disaster     USA         Florida   M  7      &    31   
2869  1979  Sea Disaster     USA         Florida   M              56   
2893  1978  Sea Disaster     USA  South Carolina   M              39   

     Fatal (Y/N)  
935            Y  
1366           N  
2197           Y  
2869           N  
2893           N

Los referentes a "Sea Disaster" también deberían ser eliminados ya que la mayoría son de accidentes de avión o barcos que han tenido un percance y no interesan para el análisis.

In [13]:
test_boat= sharks.drop(sharks[sharks['Type']!='Boat'].index)
test_boat.head()

Year  Type Country        Area  Sex  Age Fatal (Y/N)
29   2016  Boat     USA     Alabama  NaN  NaN           N
35   2016  Boat     USA  California    M  NaN           N
37   2016  Boat     USA  California    M  NaN           N
128  2015  Boat     USA  California    M  NaN           N
233  2015  Boat     USA     Florida    M  NaN           N

In [14]:
test_boating= sharks.drop(sharks[sharks['Type']!='Boating'].index)
test_boating.head()

Year     Type Country     Area  Sex  Age Fatal (Y/N)
4048  1955  Boating     USA  Florida  NaN  NaN           N
4178  1952  Boating     USA  Florida  NaN  NaN           N

Las variables que contienen "Boat" como las que contienen "Boating" nos valdrían y que son similares.

Debido a esto debemos cambiar el nombre de la variable "Boating" a "Boat" para una mejor visualización de los datos. Este proceso se hará en "sharks".

In [15]:
sharks['Type'] = sharks['Type'].str.replace('Boating', 'Boat')
print(set(sharks['Type']))

{'Sea Disaster', 'Unprovoked', 'Provoked', 'Boat', 'Invalid'}


Para comprobar que este procedimiento se ha hecho correctamente observamos la fila 4048 que estaba clasificada como "Boating" y ahora tiene que estar clasificada como "Boat".

In [16]:
sharks.loc[4048]

Year              1955
Type              Boat
Country            USA
Area           Florida
Sex                NaN
Age                NaN
Fatal (Y/N)          N
Name: 4048, dtype: object

Una vez realizado esto procedemos a eliminar todas los regitros que no contengan en "Type" la variable "Provoked", "Unprovoked" o "Boat".

In [17]:
sharks= sharks.drop(sharks[sharks['Type']=='Invalid'].index)
sharks= sharks.drop(sharks[sharks['Type']=='Sea Disaster'].index)
sharks['Type'].value_counts()

Unprovoked    1427
Provoked       150
Boat            40
Name: Type, dtype: int64

## 3. Column Area cleaning and manipulation

Una vez realizado la limpieza en la columna "Type" pasamos a reasignar las areas geográficas de USA en diferentes bloques y para ello dividiremos las geografías en:

    - East Coast
    - West Coast
    - US Island
    
Observamos los valores nulos que tenemos

In [18]:
null_cols = sharks.isnull().sum()
null_cols[null_cols > 0]

Area      1
Sex      64
Age     343
dtype: int64

Al tener solo un valor nulo en la columna "Area" procedemos a eliminarlo.

In [19]:
sharks =sharks.dropna(subset=['Area'])

Para poder agrupar las zonas geográficas en los tres grupos debemos observar los nombres de cada una para poder luego crear un diccionario que nos ayude a clasificar las zonas.

In [20]:
print(set(sharks['Area']))

{'Hawaii', 'Palmyra Atoll', 'Puerto Rico', 'Pennsylvania', 'Virginia', 'North & South Carolina', 'California', 'Maryland', 'Massachusetts', 'Guam', 'Mississippi', 'Connecticut', 'US Virgin Islands', 'Georgia', 'Washington', 'Kentucky', 'Johnston Atoll', 'Alaska', 'Wake Island', 'New Mexico', 'North Carolina', 'Maine', 'Alabama', 'Louisiana', 'Rhode Island', 'Florida', 'New York', 'New Jersey', 'South Carolina ', 'Oregon', 'Missouri', 'Texas', 'Delaware', 'South Carolina'}


Observamos que tenemos diversas formas de nombrar el estado de South Carolina y otras veces agrupan South Carolina y North Carolina en el mismo grupo. Para una mayor facilidad reemplazaremos el nombre de estas entradas en la columna "Area" por el de "North & South Carolina".

In [21]:
sharks['Area'] = sharks['Area'].str.replace('South Carolina ', 'South Carolina').str.replace('North & South Carolina', 'North Carolina')

In [22]:
print(set(sharks['Area']))
len(set(sharks['Area']))

{'Hawaii', 'Palmyra Atoll', 'Puerto Rico', 'Pennsylvania', 'Virginia', 'California', 'Maryland', 'Massachusetts', 'Guam', 'Mississippi', 'Connecticut', 'US Virgin Islands', 'Georgia', 'Washington', 'Kentucky', 'Johnston Atoll', 'Alaska', 'Wake Island', 'New Mexico', 'North Carolina', 'Maine', 'Alabama', 'Louisiana', 'Rhode Island', 'Florida', 'New York', 'New Jersey', 'Oregon', 'Missouri', 'Texas', 'Delaware', 'South Carolina'}


32

Creamos un diccionario con la tipología que representa cada área y lo agrupamos en otra columna llamada "US zone"

In [23]:
Coast = {'Washington': 'West coast', 'California': 'West coast', 'New Mexico': 'West coast', 
        'Oregon': 'West coast', 'Alaska': 'West coast', 'Maryland': 'East coast', 
        'Pennsylvania': 'East coast', 'Virginia': 'East coast', 'Rhode Island': 'East coast', 
         'Missouri': 'East coast', 'Kentucky': 'East coast', 'Georgia': 'East coast',
         'New Jersey': 'East coast', 'Delaware': 'East coast', 'Alabama': 'East coast', 
         'New York': 'East coast', 'North Carolina': 'East coast', 'Massachusetts': 'East coast', 
         'Maine': 'East coast', 'Texas': 'East coast', 'Louisiana': 'East coast', 
         'Connecticut': 'East coast', 'South Carolina': 'East coast', 'North Carolina': 'East coast', 
         'Florida': 'East coast', 'Mississippi': 'East coast', 'US Virgin Islands': 'US Islands', 
         'Guam': 'US Islands', 'Johnston Atoll': 'US Islands', 'Wake Island': 'US Islands', 
         'Puerto Rico': 'US Islands', 'Hawaii': 'US Islands', 'Palmyra Atoll': 'US Islands'}

len(Coast)


32

In [24]:
sharks['US zone']=sharks['Area'].map(Coast)
sharks.head()

Year        Type Country     Area Sex  Age Fatal (Y/N)     US zone
0  2016  Unprovoked     USA  Florida   M   16           N  East coast
1  2016  Unprovoked     USA  Florida   M   36           N  East coast
2  2016  Unprovoked     USA  Florida   M   43           N  East coast
6  2016  Unprovoked     USA  Florida   M  60s           N  East coast
7  2016  Unprovoked     USA   Hawaii   F   51           N  US Islands

In [25]:
sharks['Area'].value_counts()

Florida              852
California           223
Hawaii               198
South Carolina        98
North Carolina        72
Texas                 50
Oregon                27
Virginia              12
Alabama               11
New Jersey            10
Georgia                8
Puerto Rico            7
Massachusetts          7
Louisiana              7
New York               5
US Virgin Islands      4
Delaware               4
Mississippi            3
Maryland               3
Rhode Island           2
Palmyra Atoll          2
Alaska                 1
Guam                   1
Washington             1
Maine                  1
New Mexico             1
Wake Island            1
Johnston Atoll         1
Missouri               1
Pennsylvania           1
Connecticut            1
Kentucky               1
Name: Area, dtype: int64

In [26]:
sharks['US zone'].value_counts()

East coast    1149
West coast     253
US Islands     214
Name: US zone, dtype: int64

## 4. Column Sex cleaning

En primer lugar observamos los valores de la columna que son nulos para decidir que realizamos con ello.

In [27]:
null_cols = sharks.isnull().sum()
null_cols[null_cols > 0]

Sex     64
Age    342
dtype: int64

Observamos que solamente tenemos 64 valores nulos. Aunque es una cifra pequeña y la podemos eliminar, es mejor convertir estos valores como desconocidos en vez de eliminarlos y perder la totalidad de los valores.

In [28]:
sharks[['Sex']] = sharks[['Sex']].fillna("Uknown")

Observamos los nombres de las variables y cuantos tenemos de cada tipo.

In [29]:
sharks['Sex'].value_counts()

M         1287
F          263
Uknown      64
M            1
lli          1
Name: Sex, dtype: int64

In [30]:
print(set(sharks['Sex']))

{'M', 'F', 'M ', 'lli', 'Uknown'}


Renombramos las variables que están mal como "M " con un espacio lo cual es un error al introducirla y la convertimos a "M" o "lli" que al no saber que representa la sumamos al grupo de "Uknown". 

Además de esto cambiaremos los datos de "M" y "F" por "Male" y "Female".

In [36]:
sharks['Sex'] = sharks['Sex'].str.replace('M ', 'Male').str.replace('lli', 'Uknown').str.replace('M', 'Male').str.replace('F', 'Female')
sharks['Sex'].value_counts()

Male      1288
Female     263
Uknown      65
Name: Sex, dtype: int64

In [32]:
null_cols = sharks.isnull().sum()
null_cols[null_cols > 0]

Age    342
dtype: int64

## 5. Column Age cleaning

Como en anteriores ocasiones observamos cuantos valores nulos existen de la columna "Age".

In [35]:
null_cols = sharks.isnull().sum()
null_cols[null_cols > 0]

Age    342
dtype: int64

Al igual que en el caso de la columna "Sex" merece más la pena convertir los valores a "Uknown" que eliminarlos.

In [37]:
sharks[['Age']] = sharks[['Age']].fillna("Uknown")

In [38]:
sharks['Age'].value_counts()

Uknown      342
16           59
15           56
19           56
17           53
18           52
20           47
13           46
14           44
21           38
22           38
24           37
25           37
12           34
28           30
29           29
10           29
27           26
26           25
23           24
32           23
30           22
11           21
35           20
9            20
8            18
40           18
43           18
31           17
38           16
           ... 
63            4
Teen          4
60            4
68            4
teen          4
69            3
30s           3
20s           3
71            2
70            2
62            2
9 & 12        1
21 or 26      1
66            1
87            1
mid-30s       1
75            1
64            1
65            1
3             1
50s           1
6½            1
60s           1
73            1
Elderly       1
adult         1
60's          1
              1
12 or 13      1
86            1
Name: Age, Length: 87, d